### ISSUE TITLE
[dialyzer] connect_error should include inet and ssl error reasons?

### ISSUE THREAD

**Issue by zmstone** [epgsql/epgsql#246]
Hi.

This error spec is perhaps incomplete ?
```
-type connect_error() ::
        invalid_authorization_specification
      | invalid_password
      | {unsupported_auth_method,
         kerberosV5 | crypt | scm | gss | sspi | {unknown, integer()} | {sasl, [binary()]}}
      | {sasl_server_final, any()}
      | epgsql:query_error().
```
if it's what `epgsql_cmd_connect:execut/2` returns, it should include all possible error reasons
returned from `gen_tcp:connect/4` and `ssl:connect/3`.

`ss:reason()` however is a `any()` which is a bit unfortunate,
but fortunately it's wrapped in the ` {ssl_negotiation_failed, Reason},` tuple.

**Comment by seriyps** [comment]
Hi. What's special about this kind of error is that it is returned at the same time when connection process exits with non-normal reason:

https://github.com/epgsql/epgsql/blob/95b02f29664bc793d39db3a6576b045dd8b446f7/src/commands/epgsql_cmd_connect.erl#L83

That usually means, since epgsql connection process is started with `start_link`, that the caller of `epgsql:connect` will be killed by exit signal before it will have a chance to do smth about this returned value (??). But the return value indeed can be observed in case caller has `trap_link` flag set. We have this ticket to change this behavior: https://github.com/epgsql/epgsql/issues/232, but AFAIK, haven't yet decided if we are going to implement it or not.

So, in short, I'm fine extending this spec, but then it might need to include error reasons from those lines as well:

https://github.com/epgsql/epgsql/blob/95b02f29664bc793d39db3a6576b045dd8b446f7/src/commands/epgsql_cmd_connect.erl#L98

https://github.com/epgsql/epgsql/blob/95b02f29664bc793d39db3a6576b045dd8b446f7/src/commands/epgsql_cmd_connect.erl#L138

https://github.com/epgsql/epgsql/blob/95b02f29664bc793d39db3a6576b045dd8b446f7/src/commands/epgsql_cmd_connect.erl#L141

**Comment by zmstone** [comment]
maybe make sense to expose another `epgsql:connect` API for such error union ?

**Comment by gertcuykens** [comment]
Could it be it needs to be ssl:tls_option/0 instead? https://erlang.org/doc/man/ssl.html#type-tls_option
```
Unknown types:
  ssl:ssl_option/0
```
